src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+        self.n_heads = n_heads
+        self.ffn_size = ffn_size
+        self.embedding_size = embedding_size
+        self.activation = activation
+        self.dropout = nn.Dropout(p=dropout)
+        self.norm1 = LayerNorm(self.embedding_size, eps=layer_norm_eps)
+        self.norm2 = LayerNorm(self.embedding_size, eps=layer_norm_eps)
+        self.attention = BlenderbotMultiHeadAttention(self.n_heads, self.embedding_size, dropout=attention_dropout)
+        self.ffn = BlenderbotFFN(self.embedding_size, ffn_size, relu_dropout=relu_dropout, activation=self.activation)
+
+    def forward(self, input_tensor, mask):
+        tensor_copy = input_tensor
+        attention, _ = self.attention(tensor_copy, mask=mask)
+        input_tensor = tensor_copy + self.dropout(attention)
+        input_tensor = self.norm1(input_tensor)
+        tensor_copy = input_tensor
+        input_tensor = tensor_copy + self.dropout(self.ffn(input_tensor))
+        input_tensor = self.norm2(input_tensor)
+        input_tensor *= mask.unsqueeze(-1).type_as(input_tensor)
+        return input_tensor
+
+
+class BlenderbotEncoder(nn.Module):
+    """
+    implements the encoder model of the blender bot generative model
+    """
+
+    def __init__(self,
+                 n_heads,
+                 n_layers,
+                 embedding_size,
+                 ffn_size,
+                 vocabulary_size,
+                 embedding=None,
+                 dropout=0.0,
+                 relu_dropout=0.0,
+                 attention_dropout=0.0,
+                 padding_idx=0,
+                 embedding_scale=False,
+                 learn_positional_embeddings=False,
+                 n_positions=1024,
+                 reduction_type='mean',
+                 activation='relu',
+                 n_segments=0,
+                 output_scaling=1.0):
+        super(BlenderbotEncoder, self).__init__()
+        self.n_heads = n_heads
+        self.n_layers = n_layers
+        self.embedding_size = embedding_size
+        self.ffn_size = ffn_size
+        self.vocabulary_size = vocabulary_size
+        self.dropout = nn.Dropout(p=dropout)
+        self.reduction_type = reduction_type
+        self.padding_idx = padding_idx
+        self.embedding_scale = embedding_scale
+        self.n_segments = n_segments
+        self.n_positions = n_positions
+        self.output_dim = embedding_size
+
+        if embedding is not None:
+            assert (self.embedding_size is None or self.embedding_size == embedding.weight.shape[1]), \
+                "Embedding dimension must match embedding size"
+            self.embeddings = embedding
+        else:
+            raise AssertionError("None embedding is not allowed")
+        assert (self.embedding_size % self.n_heads == 0), "Embedding size should be a multiple of n_heads"
+        assert self.padding_idx is not None
+        self.embedding = nn.Embedding(self.vocabulary_size, self.embedding_size, padding_idx=self.padding_idx)
```

Comment:
I think it's always nice to add a separate class for both word embedding and positional embedding. It makes this file much more readable. 

self.embeddings = BlenderbotEmbeddings(config) <= this class contains word embedding and positional encoding. 

See: https://github.com/huggingface/transformers/blob/e33fdc93b4ecb571dd7a8002a74789ec8bfffc09/src/transformers/modeling_bert.py#L145